silence/silence.go

```diff
@@ -544,104 +540,179 @@ func validateSilence(s *pb.Silence) error {
+// pointers reference the same memory. If Clone() is not used before
+// replacing a silence then the operation will fail.
+//
+// Set returns an error if a silence is invalid, or would exceed any of the
+// existing limits. When this happens, a silence might have an ID even though
+// the operation failed. It is safe to re-use the silence for subsequent calls
+// to Set where previous calls have failed.
+func (s *Silences) Set(sil *pb.Silence) (err error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	now := s.nowUTC()
+	if sil.StartsAt.IsZero() {
+		sil.StartsAt = now
 	}
-	return msil.Silence, true
-}
 
-func (s *Silences) setSilence(sil *pb.Silence, now time.Time, skipValidate bool) error {
-	sil.UpdatedAt = now
+	if err := validateSilence(sil); err != nil {
+		return fmt.Errorf("invalid silence: %w", err)
+	}
 
-	if !skipValidate {
-		if err := validateSilence(sil); err != nil {
-			return fmt.Errorf("silence invalid: %w", err)
+	existing, ok := s.getSilence(sil.Id)
+	// updateExisting is true iff this is an existing silence, and the update does
+	// not affect how alerts are matched or the time window in which the
+	// silence is active (excluding time extensions).
+	updateExisting := ok && canUpdate(existing, sil, now)
+	if !updateExisting {
+		// If this is a new silence, or it is replacing an existing silence,
+		// then give it an ID and make sure the start time is not in the past.
+		uid, err := uuid.NewV4()
+		if err != nil {
+			return fmt.Errorf("failed to create UUID: %w", err)
+		}
+		sil.Id = uid.String()
+		if sil.StartsAt.Before(now) {
+			sil.StartsAt = now
 		}
 	}
+	sil.UpdatedAt = now
 
-	msil := &pb.MeshSilence{
-		Silence:   sil,
-		ExpiresAt: sil.EndsAt.Add(s.retention),
-	}
-	b, err := marshalMeshSilence(msil)
-	if err != nil {
-		return err
+	// Check the limits before expiring the existing silence to prevent
+	// partial updates when limits are exceeded. Limits are checked against
+	// the mesh silence as this is the silence that is gossiped and stored
+	// on disk.
+	msil := s.toMeshSilence(sil)
+	if err := s.checkLimits(msil, updateExisting); err != nil {
+		return fmt.Errorf("exceeds limits: %w", err)
+	}
+
+	// expireExisting is true iff this is an existing silence, and the update
+	// affects how alerts are matched or the time window in which the silence
+	// is active (excluding extensions). If existing and sil reference the
+	// same memory, then the operation will fail because the new ID does not
+	// exist. This happens when the caller does not use Clone().
+	expireExisting := ok && !updateExisting && getState(existing, now) != types.SilenceStateExpired
+	if expireExisting {
+		if err := s.expire(existing.Id); err != nil {
+			return fmt.Errorf("failed to expire silence: %w", err)
+		}
 	}
 
-	// Check the limit unless the silence has been expired. This is to avoid
-	// situations where silences cannot be expired after the limit has been
-	// reduced.
+	return s.mergeAndBroadcast(msil, now)
```

Comment:
`setSilence` was renamed to `mergeAndBroadcast`. See later comments for an explanation.